### PR TITLE
chore(elasticache): enhance service and checks

### PR DESCRIPTION
**Pull request by MrCloudSec**
### Description

Enhance AWS Elasticache service and checks since the clusters in the API are either Memcached Clusters or Redis Nodes, and the Replication Groups are Redis Clusters.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py [hunk 1]
```diff
@@ -16,9 +16,12 @@ def execute(self):
...
             report.status = "PASS"
-            report.status_extended = (
-                f"Cluster {cluster.id} is not using public subnets."
-            )
+            if cluster.engine == "redis":
+                report.status_extended = (
+                    f"Elasticache Redis Node {cluster.id} is not using public subnets."
+                )
+            else:
+                report.status_extended = f"Elasticache Memcached Cluster {cluster.id} is not using public subnets."
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py:24]: Please cover this line with a test.

File: prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py [hunk 2]
```diff
@@ -30,7 +33,10 @@ def execute(self):
 
             if len(public_subnets) > 0:
                 report.status = "FAIL"
-                report.status_extended = f"Cluster {cluster.id} is using {', '.join(public_subnets)} public subnets."
+                if cluster.engine == "redis":
+                    report.status_extended = f"Elasticache Redis Node {cluster.id} is using {', '.join(public_subnets)} public subnets."
+                else:
+                    report.status_extended = f"Elasticache Memcached Cluster {cluster.id} is using {', '.join(public_subnets)} public subnets."
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py:39]: Please cover this line with a test.

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py [hunk 3]
```diff
@@ -13,17 +13,17 @@ def execute(self):
...
             ):
                 report.status = "PASS"
-                report.status_extended = f"Elasticache Cluster {repl_group.id} has automated snapshot backups enabled with retention period {repl_group.snapshot_retention} days."
+                report.status_extended = f"Elasticache Redis cache cluster {repl_group.id} has automated snapshot backups enabled with retention period {repl_group.snapshot_retention} days."
             else:
                 if repl_group.snapshot_retention > 0:
                     report.status = "FAIL"
                     report.check_metadata.Severity = "low"
-                    report.status_extended = f"Elasticache Cluster {repl_group.id} has automated snapshot backups enabled with retention period {repl_group.snapshot_retention} days. Recommended to increase the snapshot retention period to a minimum of 7 days."
+                    report.status_extended = f"Elasticache Redis cache cluster {repl_group.id} has automated snapshot backups enabled with retention period {repl_group.snapshot_retention} days. Recommended to increase the snapshot retention period to a minimum of 7 days."
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py:27]: Please cover this line with a test.

File: prowler/providers/aws/services/elasticache/elasticache_service.py [hunk 4]
```diff
@@ -75,6 +72,34 @@ def __describe_cache_subnet_groups__(self, regional_client):
...
             )
 
+    def __describe_replication_groups__(self, regional_client):
+        # Redis Clusters
+        logger.info("Elasticache - Describing Replication Groups...")
+        try:
+            for repl_group in regional_client.describe_replication_groups()[
+                "ReplicationGroups"
+            ]:
+                replication_arn = repl_group["ARN"]
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_service.py]: Add a try/except within the for loop pelase.

File: prowler/providers/aws/services/elasticache/elasticache_service.py [hunk 5]
```diff
@@ -132,7 +151,7 @@ class Cluster(BaseModel):
     id: str
     arn: str
     region: str
-    security_groups: list[str] = []
+    engine: str
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_service.py:157]: Why this change?

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py [hunk 6]
```diff
@@ -13,17 +13,17 @@ def execute(self):
             report.resource_id = repl_group.id
             report.resource_arn = repl_group.arn
             report.status = "FAIL"
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py:16]: Please include the tags.

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled.py [hunk 7]
```diff
@@ -13,10 +13,10 @@ def execute(self):
             report.resource_id = repl_group.id
             report.resource_arn = repl_group.arn
             report.status = "FAIL"
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled.py:16]: Please include the tags.

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled.py [hunk 8]
```diff
@@ -13,14 +13,10 @@ def execute(self):
             report.resource_id = repl_group.id
             report.resource_arn = repl_group.arn
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled.py:14]: Please include the tags.

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled.py [hunk 9]
```diff
@@ -13,10 +13,10 @@ def execute(self):
             report.resource_id = repl_group.id
             report.resource_arn = repl_group.arn
             report.status = "FAIL"
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled.py:16]: Please include the tags.

File: prowler/providers/aws/services/elasticache/elasticache_service.py [hunk 10]
```diff
@@ -97,30 +122,24 @@ def __list_tags_for_resource__(self):
...
-                        transit_encryption=repl_group["TransitEncryptionEnabled"],
-                        multi_az=repl_group["MultiAZ"],
+            for repl_group in self.replication_groups.values():
+                try:
+                    regional_client = self.regional_clients[repl_group.region]
+                    repl_group.tags = regional_client.list_tags_for_resource(
+                        ResourceName=repl_group.arn
+                    )["TagList"]
+                except ClientError as error:
+                    if error.response["Error"]["Code"] == "ReplicationGroupNotFound":
```
Comment (jfagoagas) [inline prowler/providers/aws/services/elasticache/elasticache_service.py]: Can you use here `ElastiCache.Client.exceptions.ReplicationGroupNotFoundFault` instead of doing the `if`? We need to start using that as we do here https://github.com/prowler-cloud/prowler/blob/169d1686d2620e64cff90bcaf0d4f5517d93473e/prowler/providers/aws/services/iam/iam_service.py#L284

File: tests/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet_test.py
Comment (jfagoagas) [inline tests/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet_test.py:1]: Please review if we are doing an assert for the tags.

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled_test.py
Comment (jfagoagas) [inline tests/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled_test.py:1]: Please review if we are doing an assert for the tags.

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled_test.py
Comment (jfagoagas) [inline tests/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled_test.py:1]: Please review if we are doing an assert for the tags.

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled_test.py
Comment (jfagoagas) [inline tests/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled_test.py:1]: Please review if we are doing an assert for the tags.

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled_test.py
Comment (jfagoagas) [inline tests/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled_test.py:1]: Please review if we are doing an assert for the tags.

**Review by jfagoagas — CHANGES_REQUESTED** [review]
Great job @sergargar, thanks for improving the accuracy of these checks!

I left some comments to include tags in resources and also to improve some exception handling.

File: prowler/providers/aws/services/elasticache/elasticache_service.py (hunk 5 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_service.py:157]: Because the security_groups attribute is not needed, and I need the engine to know if the cluster is Redis or Memcached.

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled_test.py
Comment (MrCloudSec) [reply tests/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled_test.py:1]: Done

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled_test.py
Comment (MrCloudSec) [reply tests/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled_test.py:1]: Done

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled_test.py
Comment (MrCloudSec) [reply tests/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled_test.py:1]: Done

File: tests/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled_test.py
Comment (MrCloudSec) [reply tests/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled_test.py:1]: Done

File: tests/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet_test.py
Comment (MrCloudSec) [reply tests/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet_test.py:1]: Done

File: prowler/providers/aws/services/elasticache/elasticache_service.py (hunk 10 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_service.py]: Love this!!

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled.py (hunk 9 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_redis_cluster_rest_encryption_enabled/elasticache_redis_cluster_rest_encryption_enabled.py:16]: Done

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled.py (hunk 8 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_redis_cluster_multi_az_enabled/elasticache_redis_cluster_multi_az_enabled.py:14]: Done

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled.py (hunk 7 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_redis_cluster_in_transit_encryption_enabled/elasticache_redis_cluster_in_transit_encryption_enabled.py:16]: Done

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py (hunk 6 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py:16]: Done

File: prowler/providers/aws/services/elasticache/elasticache_service.py (hunk 4 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_service.py]: Done

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py (hunk 3 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py:27]: Done

File: prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py (hunk 2 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py:39]: Done

File: prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py (hunk 1 above)
Comment (MrCloudSec) [reply prowler/providers/aws/services/elasticache/elasticache_cluster_uses_public_subnet/elasticache_cluster_uses_public_subnet.py:24]: Done

**Comment by MrCloudSec** [comment]
Thanks for the review @jfagoagas , loved the ideas!

File: prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py (hunk 3 above)
Comment (jfagoagas) [reply prowler/providers/aws/services/elasticache/elasticache_redis_cluster_backup_enabled/elasticache_redis_cluster_backup_enabled.py:27]: Can you double check this please? I don't see a test for this case.

File: prowler/providers/aws/services/elasticache/elasticache_service.py (hunk 5 above)
Comment (jfagoagas) [reply prowler/providers/aws/services/elasticache/elasticache_service.py:157]: Good to know, thanks!

**Review by jfagoagas — CHANGES_REQUESTED** [review]

**Review by jfagoagas — APPROVED** [review]
Great one @sergargar 👏
